iaas/digitalocean/digitalocean.go

```diff
@@ -0,0 +1,397 @@
+			return
+		}
+	}
+	sshKeyRequestCreate := &godo.KeyCreateRequest{
+		Name:      "GOFN",
+		PublicKey: string(content),
+	}
+	sshKey, _, err = do.client.Keys.Create(sshKeyRequestCreate)
+	if err != nil {
+		return
+	}
+	return
+}
+
+// DeleteMachine Shutdown and Delete a droplet
+func (do *Digitalocean) DeleteMachine(machine *iaas.Machine) (err error) {
+	id, _ := strconv.Atoi(machine.ID)
+	err = do.Auth()
+	if err != nil {
+		return
+	}
+	action, _, err := do.client.DropletActions.Shutdown(id)
+	if err != nil {
+		// Power off force Shutdown
+		action, _, err = do.client.DropletActions.PowerOff(id)
+		if err != nil {
+			return
+		}
+	}
+	timeout := 120
+	quit := make(chan struct{})
+	errs := make(chan error, 1)
+	ac := make(chan *godo.Action, 1)
+	go func() {
+		for {
+			//rodando shutdown...
+			select {
+			case <-quit:
+				return
+			default:
+				d, _, err := do.client.DropletActions.Get(id, action.ID)
+				if err != nil {
+					errs <- err
+					return
+				}
+				if d.Status == "completed" {
+					ac <- d
+					return
+				}
+			}
+		}
+	}()
+	select {
+	case action = <-ac:
+		_, err = do.client.Droplets.Delete(id)
+		return
+	case err = <-errs:
+		return
+	case <-time.After(time.Duration(timeout) * time.Second):
+		err = errors.New("timed out waiting for Snhutdown")
+		return
+	}
+}
+
+// CreateSnapshot Create a snapshot from the machine
+func (do *Digitalocean) CreateSnapshot(machine *iaas.Machine) (err error) {
+	id, _ := strconv.Atoi(machine.ID)
+	err = do.Auth()
+	if err != nil {
+		return
+	}
+	action, _, err := do.client.DropletActions.Snapshot(id, "GOFN")
+	if err != nil {
+		return
+	}
+	timeout := 600
+	quit := make(chan struct{})
+	errs := make(chan error, 1)
+	ac := make(chan *godo.Action, 1)
+	go func() {
+		for {
+			//"rodando snapshot..."
+			select {
+			case <-quit:
+				return
+			default:
+				d, _, err := do.client.DropletActions.Get(id, action.ID)
+				if err != nil {
+					errs <- err
+					return
+				}
+				if d.Status == "completed" {
+					ac <- d
+					return
+				}
+			}
+		}
+	}()
+	select {
+	case action = <-ac:
+		return
+	case err = <-errs:
+		return
+	case <-time.After(time.Duration(timeout) * time.Second):
+		err = errors.New("timed out waiting for Snapshot")
+		return
+	}
+}
+
+func publicKeyFile(file string) ssh.AuthMethod {
+	buffer, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil
+	}
+	key, err := ssh.ParsePrivateKey(buffer)
+	if err != nil {
+		return nil
+	}
+	return ssh.PublicKeys(key)
+}
+
+// ExecCommand on droplet
+func (do *Digitalocean) ExecCommand(machine *iaas.Machine, cmd string) (output []byte, err error) {
+	pkPath := os.Getenv("GOFN_SSH_PRIVATEKEY_PATH")
+	if pkPath == "" {
+		pkPath = filepath.Join(keysDir, privateKeyName)
+	}
+	sshConfig := &ssh.ClientConfig{
+		User: "root",
+		Auth: []ssh.AuthMethod{
+			publicKeyFile(pkPath),
+		},
+		Timeout: time.Duration(5) * time.Minute,
+	}
+	conn, err := net.Dial("tcp", machine.IP+sshPort)
+	for err != nil {
```

Comment:
This loop might cause cpu and network saturation without a back-off or at least, a simple pause.

I suggest a pause (`time.Sleep`) of 250 ms after each try and a limit of 5 seconds (20 tries) total.

Eventually, in case of error, it should be logged on stderr (`fmt.Fprintln(os.Stderr, err`) instead of stdout (`fmt.Println(err)`) so it can be monitored and the errors catched when things go bad.